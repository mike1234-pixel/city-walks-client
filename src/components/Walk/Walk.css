/** @format */

.walk-heading-container {
  width: 100%;
  text-align: center;
}

.walk-heading {
  margin: 20px 0;
}

.author-info-container {
  width: 100%;
  text-align: center;
}

.social-links {
  width: 30%;
  display: flex;
  justify-content: space-around;
  font-size: 30px;
  margin: auto;
  margin-bottom: 20px;
}

#see-map-btn {
  position: fixed;
  bottom: 0;
  right: 1;
  left: 0;
  z-index: 10;
  background-color: var(--blue) !important;
  color: var(--elegant-color-dark);
  border: 2px var(--elegant-color-dark) solid;
}
