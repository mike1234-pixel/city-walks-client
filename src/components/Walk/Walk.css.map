{"version":3,"sourceRoot":"","sources":["Walk.scss"],"names":[],"mappings":"AAAA;AAEA;EACE;EACA;;;AAGF;EACE;EACA;;;AAGF;EACE;EACA;;;AAGF;EACE;EACA;EACA;EACA;EACA;;AAEA;EACE;;;AAIJ;EACE;EACA;EACA;EACA;EACA;EACA;EACA;EACA;EACA;EACA;;;AAGF;EACE;EACA","file":"Walk.css"}