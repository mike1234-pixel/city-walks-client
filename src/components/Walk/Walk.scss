/** @format */

.walk-heading-container {
  width: 100%;
  text-align: left;
}

.walk-heading {
  margin: 20px 0;
  font-family: var(--logo-font);
}

.author-info-container {
  width: 100%;
  text-align: left;
}

.social-links {
  width: 30%;
  display: flex;
  justify-content: space-between;
  font-size: 30px;
  margin-bottom: 20px;

  a {
    margin-right: 10px;
  }
}

#see-map-btn {
  position: fixed;
  bottom: 0;
  right: 1;
  left: 0;
  z-index: 10;
  background: var(--linear-blue) !important;
  color: var(--dark);
  border: 2px var(--dark) solid;
  transition: all 0.2s;
  transform-origin: bottom left;
}

#see-map-btn:hover {
  transform: scale(1.2);
  color: white;
}