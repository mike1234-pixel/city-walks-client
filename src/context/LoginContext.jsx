import { createContext, useState } from "react"
import axios from "axios"
import qs from "qs"
import LoginForm from "../components/pages/LoginPage/LoginForm/LoginForm";
import RegistrationForm from "../components/pages/LoginPage/RegistrationForm/RegistrationForm"
import VerificationForm from "../components/pages/LoginPage/VerificationForm/VerificationForm";
import ResetPasswordForm from "../components/pages/LoginPage/ResetPasswordForm/ResetPasswordForm";
import ForgotPasswordForm from "../components/pages/LoginPage/ForgotPasswordForm/ForgotPasswordForm";
import Boards from "../components/pages/LoginPage/LoggedInView/Boards"

export const LoginContext = createContext();

export const LoginContextProvider = (props) => {
    const [popupVisible, setPopupVisible] = useState(false)

    const [firstName, setFirstName] = useState("")
    const [lastName, setLastName] = useState("")
    const [userId, setUserId] = useState("")
    const [registrationEmail, setRegistrationEmail] = useState("")
    const [registrationPassword, setRegistrationPassword] = useState("")
  
    const [loginEmail, setLoginEmail] = useState("")
    const [loginPassword, setLoginPassword] = useState("")
  
    // save all of this in local storage
    const [loggedIn, setLoggedIn] = useState(false)
    const [userFirstName, setUserFirstName] = useState("")
    const [userLastName, setUserLastName] = useState("")
  
    const [verificationEmail, setVerificationEmail] = useState("")

    const [resetPasswordEmail, setResetPasswordEmail] = useState("")
    const [resetPasswordOldPassword, setResetPasswordOldPassword] = useState("")
    const [resetPasswordNewPassword, setResetPasswordNewPassword] = useState("")
    const [resetPasswordConfirmNewPassword, setResetPasswordConfirmNewPassword] = useState("")

    const [forgotPasswordEmail, setForgotPasswordEmail] = useState("")

    const handlePopup = () => {
      setPopupVisible(false)
      localStorage.setItem("popupVisible", false)
    }

    // set display form on LoginPage
    const [form, setForm] = useState("loggedOutView");

    let displayForm = <Boards/>;
    if (form === "boards") {
      displayForm = <Boards/>;
    } else if (form === "loginForm") {
      displayForm = <LoginForm />;
    } else if (form === "registrationForm") {
      displayForm = <RegistrationForm />;
    } else if (form === "verificationForm") {
      displayForm = <VerificationForm />;
    } else if (form === "resetPasswordForm") {
      displayForm = <ResetPasswordForm />;
    } else if (form === "forgotPasswordForm") {
      displayForm = <ForgotPasswordForm />;
    }

    const handleChangeRegistration = (event) => {
        switch(event.target.name) {
            case "registration-fname":
              setFirstName(event.target.value)
              break;
            case "registration-lname":
              setLastName(event.target.value)
              break;
            case "registration-email":
              setRegistrationEmail(event.target.value)
              break;
            case "registration-password":
              setRegistrationPassword(event.target.value)
              break;
          } 
    }
  
    const handleSubmitRegistration = (event) => {
      console.log("handle submit triggered")
        event.preventDefault()
  
        const payload = {
            fname: firstName,
            lname: lastName,
            email: registrationEmail,
            password: registrationPassword,
          };
  
    axios
        .post("http://localhost:5000/register-user", qs.stringify(payload))
        .then((res, err) => {
          if (err) {
            console.log(err);
          } else if (res.data === "An account with this email already exists.") {
            alert("An account with this email already exists.")
          } else if (res.data === "We have sent you an email. Please verify your account by clicking the link in the mail.") {
            alert("We have sent you an email. Please verify your account by clicking the link in the email. (This code expires after 10 minutes)")
           } else {
            alert("Registration Complete")
            setLoggedIn(true)
            setFirstName("")
            setLastName("")
            setRegistrationEmail("")
            setRegistrationPassword("")
            setUserFirstName(res.data.fname)
            setUserLastName(res.data.lname)
            window.scrollTo(0, 0)
          }
        });
    }
  
    const handleChangeLogin = (event) => {
      switch(event.target.name) {
          case "login-email":
            setLoginEmail(event.target.value)
            break;
          case "login-password":
            setLoginPassword(event.target.value)
            break;
        } 
  }
  
  const handleSubmitLogin = (event) => {
    console.log("handle submit login triggered")
      event.preventDefault()
  
      const payload = {
          email: loginEmail,
          password: loginPassword
        };
  
  axios
      .post("http://localhost:5000/login-user", qs.stringify(payload))
      .then((res, err) => {
        if (err) {
          console.log(err);
        } else if (res.data === "Your account exists but is not activated. Please click 'verify account' for email verification.") {
          alert("Your account exists but is not activated. Please click 'verify account' for email verification.")
        } else if (res.data === "unsuccessful login attempt") {
          alert("Unsuccessful Login Attempt. Please Try Again.")
        } else {
          alert("Login Complete")
          setLoggedIn(true)
          setLoginEmail("")
          setLoginPassword("")
          setForm("")
          setUserId(res.data._id)
          setUserFirstName(res.data.fname)
          setUserLastName(res.data.lname)
          localStorage.setItem("loggedIn", true)
          localStorage.setItem("userId", res.data._id)
          localStorage.setItem("userFirstName", res.data.fname)
          window.scrollTo(0, 0)
        }
      });
  }

  // resend verification email

  const handleChangeVerification = (event) => {
    switch(event.target.name) {
        case "verification-email":
          setVerificationEmail(event.target.value)
          break;
      }   
  }

const handleSubmitVerification = (event) => {
    console.log("handle submit verification")
    event.preventDefault()

    const payload = {
        email: verificationEmail,
      };

axios
    .post("http://localhost:5000/reverify-user", qs.stringify(payload))
    .then((res, err) => {
      if (err) {
        console.log(err);
      } else {
        alert("Verification email sent. Check your inbox.")
        setVerificationEmail("")
        window.scrollTo(0, 0)
      }
    })
}

  // reset password

  const handleChangeResetPassword = (event) => {
    switch(event.target.name) {
        case "reset-email":
          setResetPasswordEmail(event.target.value)
          break;
        case "old-password":
          setResetPasswordOldPassword(event.target.value)
          break;
        case "new-password":
          setResetPasswordNewPassword(event.target.value)
          break;
        case "confirm-new-password":
          setResetPasswordConfirmNewPassword(event.target.value)
          break;
      }   
  }

const handleSubmitResetPassword = (event) => {
    console.log("handle submit reset password")
    event.preventDefault()

    const payload = {
        email: resetPasswordEmail,
        oldPassword: resetPasswordOldPassword,
        newPassword: resetPasswordNewPassword,
      };

      if (resetPasswordNewPassword === resetPasswordConfirmNewPassword) {
        axios
        .post("http://localhost:5000/reset-password-with-old-password", qs.stringify(payload))
        .then((res, err) => {
          if (err) {
            console.log(err);
          } else if (res.data === "user not found") {
            alert("We could not find your account. Please try again.")
          } else if (res.data === "old password does not match password in the database") {
            alert("The password you entered does not match the account. Please try again or use the Forgot Password option.")
          } else if (res.data === "password successfully updated") {
            alert("Password reset. You can now login.")
            setResetPasswordEmail("")
            setResetPasswordOldPassword("")
            setResetPasswordNewPassword("")
            setResetPasswordConfirmNewPassword("")
            window.scrollTo(0, 0)
          }
        })
      } else {
        alert("Passwords don't match. Please try again.")
      }

}

  // forgot password

  const handleChangeForgotPassword = (event) => {
    switch(event.target.name) {
        case "forgot-password-email":
          setForgotPasswordEmail(event.target.value)
          break;
      }   
  }

const handleSubmitForgotPassword = (event) => {
    console.log("handle submit forgot password")
    event.preventDefault()

    const payload = {
        email: forgotPasswordEmail,
      };

axios
    .post("http://localhost:5000/forgot-password", qs.stringify(payload))
    .then((res, err) => {
      if (err) {
        console.log(err);
      } else {
        alert("We have sent you an email. Please click the click in your email to reset your password")
        setForgotPasswordEmail("")
        window.scrollTo(0, 0)
      }
    })
}
  
  const logOut = () => {
    localStorage.clear()
    localStorage.setItem("popupVisible", false)
    setLoggedIn(false)
    setUserId("")
    setUserFirstName("")
    setUserLastName("")
    setForm("boards")
  }

    return (
        <LoginContext.Provider 
            value={{
              popupVisible,
              setPopupVisible,
              handlePopup,
               // display form
               displayForm,
               setForm,
               // logged in state 
               loggedIn,
               // registration state and functions
                firstName,
                lastName,
                registrationEmail,
                registrationPassword,
                handleChangeRegistration,
                handleSubmitRegistration,
                // login state and functions
                loginEmail,
                loginPassword,
                handleChangeLogin: handleChangeLogin,
                handleSubmitLogin: handleSubmitLogin,
                setLoggedIn,
                setUserFirstName,
                setUserId,
                // logged in user data
                userFirstName,
                userId,
                userLastName,
                // resend verification email
                verificationEmail,
                handleChangeVerification,
                handleSubmitVerification,
                // reset password
                resetPasswordEmail,
                resetPasswordOldPassword,
                resetPasswordNewPassword,
                resetPasswordConfirmNewPassword,
                handleChangeResetPassword,
                handleSubmitResetPassword,
                // forgot password
                forgotPasswordEmail,
                handleChangeForgotPassword,
                handleSubmitForgotPassword,
                // logout function
                logOut
            }}
        >
            {props.children}
        </LoginContext.Provider>
    )
}

// // setter
// localStorage.setItem('myData', data);
 
// // getter
// localStorage.getItem('myData');
 
// // remove
// localStorage.removeItem('myData');
 
// // remove all
// localStorage.clear();

