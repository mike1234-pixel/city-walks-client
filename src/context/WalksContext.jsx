import { createContext, useState } from "react"

export const WalksContext = createContext();

export const WalksContextProvider = (props) => {

    const [walks, setWalks] = useState([])
    const [isLoading, setIsLoading] = useState(true)

    return (
        <WalksContext.Provider 
            value={{
                walks,
                setWalks,
                isLoading,
                setIsLoading
            }}
        >
            {props.children}
        </WalksContext.Provider>
    )
}