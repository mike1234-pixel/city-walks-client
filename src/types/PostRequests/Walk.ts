export default interface Walk {
    walk: string;
    city: string;
    description: string;
    startingPoint: string;
    content1: string;
    content2: string;
    content3: string;
    coverImg: string;
    mapImg: string;
    img1: string;
    img2: string;
    img3: string;
    author: string;
    aboutTheAuthor: string;
    websiteLink: string;
    instagramLink: string;
    facebookLink: string;
    twitterLink: string;
    lat: string;
    lng: string;
    length: string;
    iframeLink: string;
    iframeTitle: string;
}