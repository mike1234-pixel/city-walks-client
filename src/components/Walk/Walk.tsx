import React, { ReactNode, useEffect, useState } from "react";
import { MDBIcon, MDBAnimation, MDBBtn, MDBContainer } from "mdbreact";
import SectionA from "../SectionA/SectionA";
import SectionB from "../SectionB/SectionB";
import PopUp from "./PopUp/PopUp";
import LoadingBar from "../LoadingBar/LoadingBar";
import toTitleCase from "../../functions/toTitleCase";
import { BsDot } from "react-icons/bs";
import { FaRoad, FaMapMarkerAlt } from "react-icons/fa";
import { IoMdTrain } from "react-icons/io";
import { connect } from "react-redux";
import { History } from "history";
import RootState from "../../types/State/Root/State";
import WalkI from "../../types/Walks/Walk";
import "./Walk.css";

interface Props {
  walks: Array<WalkI>;
  history: History;
}

const Walk: React.FC<Props> = (props: Props) => {
  const { walks, history } = props;

  const [togglePopUp, setTogglePopUp] = useState<boolean>(false);

  const walkName: string = toTitleCase(
    history.location.pathname.replace("/walks/", "").replace(/-/g, " ")
  );

  useEffect(() => {
    window.scrollTo(0, 0);
    document.addEventListener("mousedown", handleClickOutside);
  });

  let walk: string | ReactNode = <LoadingBar />;

  let currentWalk: WalkI | { iframeLink: string; iframeTitle: string; } = {
    iframeLink: "",
    iframeTitle: "",
  };

  const handleClick: () => void = () => {
    setTogglePopUp(!togglePopUp);
  };

  // THIS EVENT SHOULD HAVE A MORE SPECIFIC TYPE
  const handleClickOutside: (e: any) => void = (e) => {
    if (
      e.target.id !== "popup-img" &&
      e.target.id !== "see-map-btn" &&
      e.target.id !== "map-icon"
    ) {
      setTogglePopUp(false);
    }
  };

  if (walks.length) {
    let selectedWalk: Array<WalkI> | WalkI | undefined = walks.filter(
      (walk) => walk.walk === walkName
    );
    selectedWalk = selectedWalk[0];

    currentWalk = selectedWalk;

    if (!selectedWalk) {
      walk = "walk not found";
    } else {
      walk = (
        <div>
          <div className='walk-heading-container'>
            <h1 className='walk-heading'>
              {selectedWalk.walk} <BsDot /> {selectedWalk.city}
            </h1>
            <p className='walk-description'>
              <FaMapMarkerAlt /> {selectedWalk.description}
            </p>
            <p className='walk-description'>
              <IoMdTrain /> Starting Point: {selectedWalk.startingPoint}
            </p>
            <p className='walk-description'>
              <FaRoad /> Length: {selectedWalk.length}
            </p>
          </div>
          <SectionA
            content={selectedWalk.content1}
            img={selectedWalk.img1}
            alt={selectedWalk.walk}
            zoom={false}
            waves={false}
          />
          <SectionB
            content={selectedWalk.content2}
            img={selectedWalk.img2}
            alt={selectedWalk.walk}
            zoom={false}
            waves={false}
          />
          <SectionA
            content={selectedWalk.content3}
            img={selectedWalk.img3}
            alt={selectedWalk.walk}
            zoom={false}
            waves={false}
          />
          <div className='author-info-container'>
            <p>This walk was written by {selectedWalk.author}</p>
            <p>{selectedWalk.aboutTheAuthor}</p>
            <p>Connect with {selectedWalk.author.split(" ")[0]}!</p>
            <MDBAnimation reveal type='rubberBand'>
              <div className='social-links'>
                {selectedWalk.websiteLink !== undefined && (
                  <a href={selectedWalk.websiteLink} target='_blank'>
                    <MDBIcon fab icon='github' />
                  </a>
                )}
              </div>
            </MDBAnimation>
          </div>
        </div>
      );
    }
  }

  return (
    <div>
      <MDBContainer>
        <div data-testid='walk-page'>
          <div className='page center'>{walk}</div>
        </div>
      </MDBContainer>
      {!togglePopUp && (
        <MDBBtn id='see-map-btn' onClick={handleClick}>
          See Map <MDBIcon icon='map-marked-alt' id='map-icon' />
        </MDBBtn>
      )}
      {togglePopUp && (
        <PopUp
          iframeLink={currentWalk.iframeLink}
          iframeTitle={currentWalk.iframeTitle}
        />
      )}
    </div>
  );
};

const mapStateToProps: (state: RootState) => void = (state) => ({
  walks: state.walksState.walks,
});

export default connect(mapStateToProps)(Walk);
