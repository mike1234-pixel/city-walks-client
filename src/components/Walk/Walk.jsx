import { useEffect, useState, useContext } from "react"
import { MDBIcon, MDBAnimation, MDBBtn } from "mdbreact"
import SectionA from '../SectionA/SectionA'
import SectionB from '../SectionB/SectionB'
import PopUp from './PopUp/PopUp'
import toTitleCase from "../../functions/toTitleCase"
import { WalksContext } from "../../context/WalksContext"
import PrivacyPopUp from "../../components/PopUp/PopUp"
import { LoginContext } from "../../context/LoginContext"
import './Walk.css'

const Walk = ({match}) => {

  const [togglePopUp, setTogglePopUp] = useState(false)

  const handleClick = () => {
    setTogglePopUp(!togglePopUp)
  }

  const { walks, isLoading } = useContext(WalksContext)
  const { popupVisible } = useContext(LoginContext)

  const walkName = toTitleCase(match.url.replace("/walks/", "").replace(/-/g, " "))

    useEffect(() => {
        window.scrollTo(0, 0);
      });

      let walk = "loading";

      if (!isLoading) {

        let selectedWalk = walks.filter((walk) => walk.walk === walkName)
        selectedWalk = selectedWalk[0]

        if (selectedWalk === undefined) {
          walk = "walk not found"
        } else {
            walk = 
            <div>
            <div className="walk-heading-container">
              <h1 className="walk-heading display-font">{`${selectedWalk.walk} -- ${selectedWalk.city}`}</h1>
              <p>{selectedWalk.description}</p>
              <p>Starting Point: {selectedWalk.startingPoint}</p>
            </div>
              <SectionA content={selectedWalk.content1} img={selectedWalk.img1} alt={selectedWalk.walk}/>
              <SectionB content={selectedWalk.content2} img={selectedWalk.img2} alt={selectedWalk.walk}/>
              <SectionA content={selectedWalk.content3} img={selectedWalk.img3} alt={selectedWalk.walk}/>
              <div className="author-info-container">
                <p>This walk was written by {selectedWalk.author}</p>
                <p>{selectedWalk.aboutTheAuthor}</p>
                <p>Connect with {selectedWalk.author.split(" ")[0]}!</p>
                <MDBAnimation reveal type="rubberBand">
                  <div className="social-links">
                    {selectedWalk.websiteLink !== undefined && <a href={selectedWalk.websiteLink} target="_blank"><MDBIcon icon="laptop" /></a>}
                    {selectedWalk.facebookLink !== undefined && <a href={selectedWalk.facebookLink} target="_blank"><MDBIcon fab icon="facebook" /></a>}
                    {selectedWalk.instagramLink !== undefined && <a href={selectedWalk.instagramLink} target="_blank"><MDBIcon fab icon="instagram" /></a>}
                    {selectedWalk.twitterLink !== undefined && <a href={selectedWalk.twitterLink} target="_blank"><MDBIcon fab icon="twitter" /></a>}
                  </div>
                </MDBAnimation>
                <MDBBtn id="see-map-btn" onClick={handleClick} >
                {togglePopUp ? "Unsee Map" : "See Map" } <MDBIcon icon="map-marked-alt" />
                </MDBBtn>
                {togglePopUp && <PopUp mapImg={selectedWalk.mapImg} handleClick={handleClick}/>}
              </div>
          </div>
        }
      }


    return (
      <div data-testid="walk-page">
        {popupVisible && <PrivacyPopUp/>}
        <div className="min-page-height center">
          {walk}
        </div>
      </div>
    )
}

export default Walk