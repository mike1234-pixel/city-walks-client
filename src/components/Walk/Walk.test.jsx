import React from 'react'
import renderer from 'react-test-renderer'
import { WalksContextProvider } from "../../context/WalksContext"
import { LoginContextProvider } from "../../context/LoginContext"
import Walk from './Walk'

// snapshot test
it('Walk renders correctly', () => {  

  const tree = renderer
    .create(
        <WalksContextProvider>
            <LoginContextProvider>
                <Walk match={{params: {id: 1}, isExact: true, path: "", url: "/walks/edmonton-green"}}/>
            </LoginContextProvider>
        </WalksContextProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});