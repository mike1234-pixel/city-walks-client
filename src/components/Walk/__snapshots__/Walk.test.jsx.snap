// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Walk renders correctly 1`] = `
<div
  data-testid="walk-page"
>
  <div
    className="min-page-height center"
  >
    loading
  </div>
</div>
`;
